---
layout: post
title:  "Detecting selection in protein-coding sequences"
permalink: /day1/
category: courses
---

## Overview

This session will address the question: **Given a multiple sequence alignment containing protein-coding sequences and a corresponding phylogeny, what kinds of questions can we ask about natural selection?**

We will investigate the signatures of natural selection in protein-coding sequence data usng [HyPhy](http://hyphy.org), via the [Datamonkey webserver](datamonkey.org).

> Note: Newer analyses (like BUSTED, aBSREL, and RELAX) are only available via the [development Datamonkey webserver](test.datamonkey.org).



## Background

These papers provide excellent overviews (with technical details as well) of phylogenetic codon models:

+ [Investigating Protein-Coding Sequence Evolution with Probabilistic Codon Substitution Models 
](https://doi.org/10.1093/molbev/msn232)
+ [Trends in Substitution Models of Molecular Evolution](https://dx.doi.org/10.3389/fgene.2015.00319)
+ If you want to dive in as much as possible, this book is the definitive overview of molecular evolution, including codon models: [Computational Molecular Evolution](https://www.amazon.com/Molecular-Evolution-Statistical-Ziheng-Yang/dp/0199602611/ref=pd_cp_14_1?_encoding=UTF8&pd_rd_i=0199602611&pd_rd_r=YDZH0B4YMAH1P0W9T4FQ&pd_rd_w=YqzqK&pd_rd_wg=EIazf&psc=1&refRID=YDZH0B4YMAH1P0W9T4FQ)

## Materials

You can download slides for today's session [here](../day1/selection.pptx).


### Data

We will investigate selection using two example datasets:

1. A dataset of 10 CD2 mammalian orthologs.
	+ HyPhy-formatted [dataset](../day1/cd2_hyphy_input.txt)
	+ Link to [alignment only](../day1/cd2_alignment.fasta)
	+ Link to [tree only](../day1/cd2_tree.nwk)
2. A dataset containing Borna disease virus sequences from both endogenous and closely-related "free-living" viruses. 
	+ HyPhy-formatted [dataset](../day1/borna_hyphy_input.txt)
	+ Link to [alignment only](../day1/borna_alignment.fasta)
	+ Link to [tree only](../day1/borna_tree.nwk)

> Protip: [This python script](../day1/prepare_hyphy_input.py) is a useful wrapper for prepping your data for HyPhy input. It is written in Python2 and requires that the aligner [mafft](http://mafft.cbrc.jp/alignment/software/) and the phylogenetic reconstruction method [FastTree](http://www.microbesonline.org/fasttree/) are installed and available in your path.

Note that you can use this [online widget](veg.github.io/phylotree.js) to visualize and annotate phylogenies.

### Results

Method | Input data | Datamonkey Version | Result page(s)*
-------|------------|-------------|------------
BUSTED | [CD2](../day1/cd2_hyphy_input.txt) | [test.datamonkey.org](http://test.datamonkey.org) | [Subset FG Results](http://test.datamonkey.org/busted/591c9f81c0fbab8023e7bdc3) and [All FG Results](http://test.datamonkey.org/busted/591ca1ebc0fbab8023e7bdd4)
aBSREL | [Borna](../day1/borna_hyphy_input.txt) | [test.datamonkey.org](http://test.datamonkey.org)
RELAX  | [Borna](../day1/borna_hyphy_input.txt) | [test.datamonkey.org](http://test.datamonkey.org) | [Results](http://test.datamonkey.org/relax/591c9ef7c0fbab8023e7bd98)
FEL    | [CD2](../day1/cd2_hyphy_input.txt) |  [datamonkey.org](http://datamonkey.org)
MEME   | [CD2](../day1/cd2_hyphy_input.txt) | [datamonkey.org](http://datamonkey.org)

**\*Note**: Some of these links will expire a few days after this workshop, so please re-run analyses in the future.


